File: utils/scripts/helpers/naming.ts
/**
 * Naming convention utilities for code generation
 * Converts between different naming styles: underscore_case, camelCase, PascalCase, Title Case
 */

export const RESERVED_WORDS = new Set([
  'break', 'case', 'catch', 'class', 'const', 'continue', 'debugger', 'default', 'delete',
  'do', 'else', 'enum', 'export', 'extends', 'false', 'finally', 'for', 'function',
  'if', 'implements', 'import', 'in', 'instanceof', 'interface', 'let', 'new', 'null',
  'package', 'private', 'protected', 'public', 'return', 'super', 'switch', 'static',
  'this', 'throw', 'true', 'try', 'typeof', 'var', 'void', 'while', 'with', 'yield',
  'await'
]);

/**
 * Converts underscore_case to camelCase
 * Example: "user_name" → "userName"
 */
export function toCamelCase(str: string): string {
  return str.replace(/_([a-z])/g, (_, letter) => letter.toUpperCase());
}

/**
 * Converts underscore_case to PascalCase
 * Example: "user_name" → "UserName"
 */
export function toPascalCase(str: string): string {
  const camel = toCamelCase(str);
  return camel.charAt(0).toUpperCase() + camel.slice(1);
}

/**
 * Capitalizes the first letter of a string (assumes already camelCase)
 * Example: "userName" → "UserName"
 */
export function toPascalCaseFromVar(str: string): string {
  return str.charAt(0).toUpperCase() + str.slice(1);
}

/**
 * Converts underscore_case to Title Case (for UI labels)
 * Example: "user_name" → "User Name"
 */
export function toTitleCase(str: string): string {
  return str.split('_').map(word => word.charAt(0).toUpperCase() + word.slice(1)).join(' ');
}

/**
 * Converts to camelCase and escapes JavaScript reserved words
 * Example: "class" → "classValue", "user_name" → "userName"
 */
export function safeVarName(name: string): string {
  const camel = toCamelCase(name);
  return RESERVED_WORDS.has(camel) ? `${camel}Value` : camel;
}

/**
 * Removes trailing 's' from a name (basic pluralization)
 * Preserves the original case style
 * Example: "users" → "user", "Users" → "User"
 */
export function singularize(name: string): string {
  if (!name) return name;
  const isPascal = name[0] === name[0].toUpperCase();
  const lower = name.toLowerCase();
  if (!lower.endsWith('s')) {
    return name;
  }
  const base = name.slice(0, -1);
  if (isPascal) {
    return base.charAt(0).toUpperCase() + base.slice(1);
  }
  return base;
}

File: code_generator/helpers/naming.py
"""Naming convention utilities — port of helpers/naming.ts"""
import re

RESERVED_WORDS = {
    'break', 'case', 'catch', 'class', 'const', 'continue', 'debugger', 'default', 'delete',
    'do', 'else', 'enum', 'export', 'extends', 'false', 'finally', 'for', 'function',
    'if', 'implements', 'import', 'in', 'instanceof', 'interface', 'let', 'new', 'null',
    'package', 'private', 'protected', 'public', 'return', 'super', 'switch', 'static',
    'this', 'throw', 'true', 'try', 'typeof', 'var', 'void', 'while', 'with', 'yield',
    'await',
}


def to_camel_case(s: str) -> str:
    """user_name -> userName"""
    return re.sub(r'_([a-z])', lambda m: m.group(1).upper(), s)


def to_pascal_case(s: str) -> str:
    """user_name -> UserName"""
    camel = to_camel_case(s)
    return camel[0].upper() + camel[1:] if camel else camel


def to_pascal_case_from_var(s: str) -> str:
    """camelCase -> PascalCase (capitalise first letter only, no underscore splitting)"""
    return s[0].upper() + s[1:] if s else s


def to_title_case(s: str) -> str:
    """user_name -> User Name"""
    return ' '.join(word[0].upper() + word[1:] for word in s.split('_'))


def safe_var_name(name: str) -> str:
    """delete -> deleteValue, user_name -> userName"""
    camel = to_camel_case(name)
    return f'{camel}Value' if camel in RESERVED_WORDS else camel


def singularize(name: str) -> str:
    """shift_assignments -> shift_assignment (preserves case style)"""
    if not name:
        return name
    is_pascal = name[0] == name[0].upper()
    if not name.lower().endswith('s'):
        return name
    base = name[:-1]
    if is_pascal:
        return base[0].upper() + base[1:] if base else base
    return base
